symfony/framework-bundle/4.2/src/bootstrap.php

```diff
@@ -0,0 +1,11 @@
+<?php
+
+require __DIR__.'/../vendor/autoload.php';
+
+use Symfony\Component\Dotenv\Dotenv;
+
+if (class_exists(Dotenv::class)) {
+    (new Dotenv())->loadForEnv($_SERVER['APP_ENV'] ?? $_ENV['APP_ENV'] ?? 'dev', __DIR__.'/../.env');
```

Comment:
This change will always load the .env if I understand it correctly. Now if I make a mistake in composer.json or a composer command where I forget `--dev`, it will move `Dotenv` from `require-dev` to `require` and suddenly my production server is using the wrong variables. The same goes for installation when missing `--no-dev`, you will deploy _with_ Dotenv and break your production, possibly connecting to the wrong database.

I have not followed all discussions, I came here from https://github.com/symfony/symfony/issues/25693#issuecomment-424329159

But this sounds like a very dangerous change that has the potential to break things _by default_. You have to change something in order to _not_ load the wrong data (`--no-dev`) and it's unforgiving when you make a mistaken with your dependencies. Both of those which I see happen all the time with developers new to composer.

Reply:
@iltar yes it's 100% intended (see the proposal of @ogizanagi in the old comments of this PR).

It allows user that can't (or don't want to) properly setup environment variables to use `.env.local` as they was using `parameters.yml` with Symfony 3, while still following (by default and by design) 12factor methodology/cloud native.

It also fixes a lot of usability problems discussed here.

Reply:
I get that this is the intention, but it would make it really easy to critically break production environments if you're not paying attention, which sounds too much of a risk for the possible gain.

To me that sounds like a no-go.

Reply:
@iltar If the env var is already set, DotEnv won't override it anyway, maybe this could comfort you a bit :D

Reply:
By default, the skeleton has [`symfony/dotenv` under the require-dev section](https://github.com/symfony/skeleton/blob/master/composer.json).
The [documentation](https://symfony.com/doc/current/components/dotenv.html) mentions clearly the Dotenv component is not recommended to be used on production. But we must probably update the installation section to add the `--dev` option.

Worst case indeed is the wrong env value will be used if the Dotenv is installed on production, but only if the env var is actually missing on the server. That should be prevented by learning developers new to composer how to take advantage of its features and ensuring your deployment properly uses the `--no-dev` option. When deploying you could also remove the `.env` file for safety. 
IMHO that shouldn't prevent us from doing the changes in this PR.

However, if anyone else is reluctant to these changes as is, we can just add an extra `'prod' !== $env` condition before actually calling `Dotenv:: loadForEnv()`, as a tradeoff. Developers knowing what it implied can still remove it easily.

Reply:
Maybe add `\trigger_error('You should not use Dotenv in production', \E_USER_WARNING);` here or inthe Dotenv component to warn users ?

Reply:
One of the goal is to allow users to use Dotenv in production (even if it's not recommended, it's not risked, it's just slower and not a good practice). I don't think we should add this warning. We should just document it properly.

And as pointed by @Pierstoval and @ogizanagi there are no security implication at all because if an env var is already set, it will never be overrode.

Reply:
There is still some risk here: Some code may only be executed if a certain environment variable is set. And the Dotenv component may not be installed accidentally by the developer themself, but because some third-party package has it in its `require` section.

Reply:
I'm not sure to understand, can you give an example?

> some third-party package has it in its require section

It would be a super bad practice.

Reply:
I agree. I can't see any good reason for a third-party bundle to require the Dotenv component. The component targets end-projects.

Reply:
We are all developers and we know that people apply bad practices all the time. Sometimes developers need to be protected and this is one of those scenarios where the little bit of protection can prevent a catastrophic failure.

Reply:
How can it be catastrophic? Existing env vars are never overrode.

Reply:
If I have no APP_ENV set (or any others), and the package ended up in the require, in this example it will try to load the DEV env. If for some reason the development database etc can be reached (which isn't unlikely), or a development micro-service, all requests in production will start talking to this environment. This can lead to incorrect payments, sensitive data leaking because of user ids being different and for some reason test data is using semi-real production data etc.

In this scenario, I rather have my application crash into a 500 than having it connect to the wrong services. By making it the default behavior, you'll open up a can of worms that can harm companies by making an innocent mistake. I'd rather see the development fail on first try, than production, even if this hurts DX in the process. 

```php
(new Dotenv())->loadForEnv($_SERVER['APP_ENV'] ?? $_ENV['APP_ENV'] ?? 'dev', __DIR__.'/../.env')
```

Reply:
I also feel loading env file just because DotEnv is installed is too risky as the default. The compromise suggested by @ogizanagi would be my preferred solution.

> However, if anyone else is reluctant to these changes as is, we can just add an extra 'prod' !== $env condition before actually calling Dotenv:: loadForEnv(), as a tradeoff. Developers knowing what it implies can still remove it easily.

Reply:
done